### DISCUSSION TITLE
Number of Clients

### DISCUSSION THREAD

**Discussion by dadams39** [microsoftgraph/msgraph-sdk-go#200] [Q&A] [answered]
I distinctly remember a thread a thread on GraphServiceClients where it stated that it was not advisable to make more than one for a program. Does anyone know what the maximum number of clients is and where that is documented?

**Comment by baywet** [comment, marked as answer]
This simple question is difficult to answer by nature. If you look at the Graph Service Client it caries multiple things:
- The entry point to the chained method API surface. This is just a large tree, although we wouldn't recommend creating thousands of those (there's no point in doing so), your application should be fine as long as it doesn't run out of memory.
- The serialization/deserilization factories. They are lightweight and shouldn't be a source of trouble.
- The http request adapter. This carries the authentication adapter and the http client. Both of these carry connections pools. The request adapter will instantiate a new HTTP client is none is passed. And this kind of behaviour can rapidly lead to port exhaustion.

The main reason to have to instantiate multiple request adapters is usually to swap authentication context (i.e. on behalf flow or similar, with lots of different users). In which case our advice it to create an http client using the factories provided in the core package, keep it as a singleton, and pass it when instantiating request adapters to be used with authentication providers/service clients as needed.

The bottom line being:
- http clients are expensive
- http request adapters are cheap
- authentication providers can be expensive since they also rely on a client, but you can reach out to azure identity to know more about reusing the http client and make them cheap as well.
- serialization factories are cheap
- service clients are moderately cheap (large tree)

[more reading on port exhaustion in the content of Go](https://www.sobyte.net/post/2022-03/go-http-client-connection-control/)
